### ISSUE TITLE
Revisit "Declare variables at the top of their block scope"

### ISSUE THREAD

**Issue by brzpegasus** [dockyard/styleguides#46]
In our JavaScript style guide, we have this rule: **"Declare variables at the top of their block scope".**

I'd like to open this issue so we can talk about the rationale behind it, as I understand why it's best practice with `var`, but not so much now that we've switched to `const` or `let`.

For example, we do this a lot in projects:

```js
  actions: {
    submit(callback) {
      if (get(this, 'isInvalid')) {
        callback(RSVP.reject());
        set(this, 'shouldShowValidationErrors', true);
        return;
      }

      const service = get(this, 'newsletterService');
      const data = this.$('form').serialize();

      const promise = service.signup(data)
        .then(() => {
          // do something on success
        }, (error) => {
          // do something on error
        });

      callback(promise);
    }
  }
```

If we were to require declaring variables at the top, we would need to change the code to:

```js
  actions: {
    submit(callback) {
      // declare (and initialize here), even if we don't care to access the service,
      // or to serialize the form data if the form is invalid.
      const service = get(this, 'newsletterService');
      const data = this.$('form').serialize();

      let promise; // now this becomes a `let` since we don't want to assign a value until later

      if (get(this, 'isInvalid')) {
        callback(RSVP.reject());
        set(this, 'shouldShowValidationErrors', true);
        return;
      }

      promise = service.signup(data)
        .then(() => {
          // do something on success
        }, (error) => {
          // do something on error
        });

      callback(promise);
    }
  }
```

Alternatively, we could do this:

```js
  actions: {
    submit(callback) {
      if (get(this, 'isInvalid')) {
        callback(RSVP.reject());
        set(this, 'shouldShowValidationErrors', true);
      } else {
        const service = get(this, 'newsletterService');
        const data = this.$('form').serialize();

        const promise = service.signup(data)
          .then(() => {
            // do something on success
          }, (error) => {
            // do something on error
          });

        callback(promise);
      }
    }
  }
```

But I'm generally a fan of early return statements, especially if the `if`/`else` blocks aren't short enough.

Most importantly, I'm wondering if hoisting variables to the top is still necessary, and if yes, why it's dangerous not to do so. With `var`, trying to access a variable before it is declared gives you `undefined` because the declaration is hoisted to the top of the block. No error is raised, leading to potentially hard to find bugs. With `const` and `let`, doing so causes a `ReferenceError`, so it's a lot more obvious.

**Comment by rwjblue** [comment]
IMHO, the rule only makes sense with 'var'.

**Comment by bcardarella** [comment]
Agreed. There is also something to be said about allocating memory unnecessarily. If we're declaring variables before a potential `return` statement this is not ideal, even if it is trivial.

**Comment by poteto** [comment]
For some context, the original function I commented on (that sparked the discussion) looked like this:

```js
  actions: {
    submit(callback) {
      const data = get(this, 'data');
  
      if (get(data, 'isInvalid')) {
        set(this, 'shouldShowValidationErrors', true);
        callback(reject());
        return;
      }
  
      data.applyChanges();
  
      const account = get(this, 'account');
      const { year, month, day } = getProperties(data, 'year', 'month', 'day');
      set(account, 'dateOfBirth', foo(year, month, date));
  
      callback(account.save());
    }
  }
```

My comment was to move the 2 `const`s (account and the destructured year, month, day) below `const data = get(this, 'data');`, as per the rule in question, and because TDZ did not come into effect. Because of the new `let` and `const` semantics  (see [ES2015 spec](http://www.ecma-international.org/ecma-262/6.0/#sec-let-and-const-declarations)), `let` and `const` do still hoist variables to the top of the block scope, but throw ReferenceErrors if accessed prior to initialization.

So the argument re: allocating memory unnecessarily isn't quite valid, as they are still being hoisted and allocated regardless.

Personally, I think it reads better if we apply the rule, but we should probably add scenarios to the styleguide where it is OK to break the rule.

**Comment by poteto** [comment]
I think the examples above are a tiny bit contrived. How about this example:

```js
// original
  actions: {
    submit(callback) {
      const data = get(this, 'data');
  
      if (get(data, 'isInvalid')) {
        // do stuff
        return;
      }
  
      const thing = get(this, 'thing');
      const anotherThing = getProperties(data, 'year', 'month', 'day');
      set(thing, 'dateOfBirth', anotherThing);
  
      callback(thing.save());
    }
  }
```

With the rule in effect, we declare variables at the top:

```js
// with the current rule applied
actions: {
  submit(callback) {
    const data = get(this, 'data');
    const thing = get(this, 'thing');
    const anotherThing = getProperties(data, 'year', 'month', 'day');

    if (get(data, 'isInvalid')) {
      // do stuff
      return;
    }

    set(thing, 'dateOfBirth', anotherThing);
    callback(account.save());
  }
}
```

Because of the new `let` and `const` semantics  (see [ES2015 spec](http://www.ecma-international.org/ecma-262/6.0/#sec-let-and-const-declarations)), `let` and `const` do still hoist variables to the top of the block scope, but throw ReferenceErrors if accessed prior to initialization:

>The variables are **created when their containing Lexical Environment is instantiated** but may not be accessed in any way until the variable’s LexicalBinding is evaluated. 

So the argument re: allocating memory unnecessarily isn't quite valid, as they are still being hoisted and allocated regardless (correct me if I'm wrong).

Personally, I think it reads better (by grouping variable declarations together when possible) if we continue to apply the rule, but we should add scenarios to the style guide describing where it is OK to break the rule, for example, where TDZ applies, or in the case of early returns:

```js
// group variable declarations at the top, unless returning early
// in which case they should be declared together
actions: {
  submit(callback) {
    if (get(this, 'data.isInvalid')) {
      // do stuff
      return;
    }
    
    const data = get(this, 'data');
    const thing = get(this, 'thing');
    const anotherThing = getProperties(data, 'year', 'month', 'day');

    set(thing, 'dateOfBirth', anotherThing);
    callback(account.save());
  }
}
```

Thoughts?

**Comment by brzpegasus** [comment]
The variables get created when their containing lexical environment is instantiated, but it doesn't mean the right hand assignments would also get evaluated then. I do think it's important not to make function calls unless you need them, and in the case of early returns, it is best to delay.

Also, the fact that an explicit `ReferenceError` gets thrown if a variable is accessed before it is declared (one of the differentiating points between `let`/`const` and `var`) makes the whole argument about needing to hoist to the top moot in my opinion.

**Comment by poteto** [comment]
Agreed – I don't feel strongly about it, but as I mentioned in my earlier comment, I think it reads better (see last example), so I think we should add scenarios to the rule (or perhaps remove it altogether?) describing when it is okay to ignore.
